magento/magento2
# Include custom source model options in attribute index - Up port #16727

### Description
This is an up-port of #16727 from 2.2 to 2.3

When a multiselect Product Attribute uses a custom source model those values are not included when indexing the attribute values for filters on the product grid. The result is those attributes cannot be used as a filter in the product grid. 

This was originally opened as [issue #417](https://github.com/magento/magento2/issues/417) but was closed as it was considered a feature request and moved to the forums [here](https://community.magento.com/t5/Magento-2-x-Feature-Requests-and/Attribute-Indexing-ignores-custom-source-model-options/idi-p/46851). The solution is based on @maderlock's comment [here](https://github.com/magento/magento2/issues/417#issuecomment-265146285) but replaced the use of `ObjectManager` with dependency injection of `AttributeRepositoryInterface`.

### Fixed Issues
1. magento/magento2#417 : Attribute Indexing ignores custom source model options

### Manual testing scenarios
1. Create a multiselect attribute with a custom source model. For a quick test I created this [sample module](https://github.com/rain2o/magento2-Rain2o_MultiselectSourceSample).
2. Within the attribute set Use in Layered Navigation to "Filterable (with results)" (if using the sample module this is already done.)
3. Select one or more values for the given attribute on two or more products in the same category.
4. View the category page. The new attribute should now be an available filter.

### Contribution checklist
